infrastructure/src/main/java/org/corfudb/infrastructure/log/SegmentSupervisor.java

```diff
@@ -0,0 +1,181 @@
+            FileChannel writeCh = null;
+            FileChannel readCh = null;
+
+            try {
+                writeCh = getChannel(missingFilePath, false);
+                readCh = getChannel(missingFilePath, true);
+
+                SegmentHandle segment = new SegmentHandle(segmentId, writeCh, readCh, missingFilePath);
+                // The first time we open a file we should read to the end, to load the
+                // map of entries we already have.
+                // Once the segment address space is loaded, it should be ready to accept writes.
+                addressSpaceLoader.load(segment);
+                return segment;
+            } catch (IOException e) {
+                log.error("Error opening file {}", missingFilePath, e);
+                IOUtils.closeQuietly(writeCh);
+                IOUtils.closeQuietly(readCh);
+                throw new IllegalStateException(e);
+            } catch (RuntimeException ex) {
+                //Prevents file resources leaks in case of any RuntimeException.
+                IOUtils.closeQuietly(writeCh);
+                IOUtils.closeQuietly(readCh);
+                throw ex;
+            }
+        });
+
+        handle.retain();
+        return handle;
+    }
+
+    @Nullable
+    private FileChannel getChannel(Path filePath, boolean readOnly) throws IOException {
+        if (readOnly) {
+            if (!filePath.toFile().exists()) {
+                throw new FileNotFoundException(filePath.toString());
+            }
+
+            return FileChannel.open(filePath, EnumSet.of(StandardOpenOption.READ));
+        }
+
+        EnumSet<StandardOpenOption> options = EnumSet.of(
+                StandardOpenOption.READ,
+                StandardOpenOption.WRITE,
+                StandardOpenOption.CREATE
+        );
+        FileChannel channel = FileChannel.open(filePath, options);
+
+        // First time creating this segment file, need to sync the parent directory
+        File segFile = filePath.toFile();
+        syncDirectory(segFile.getParent());
+        return channel;
+    }
+
+    /**
+     * Close a segment
+     *
+     * @param segment segment handler
+     */
+    public void close(SegmentHandle segment) {
+        if (segment.getRefCount() != 0) {
+            log.warn("closeSegmentHandlers: Segment {} is trimmed, but refCount is {}, attempting to trim anyways",
+                    segment.getSegment(), segment.getRefCount()
+            );
+        }
+
+        channelsToSync.remove(segment.getWriteChannel());
+        channels.remove(segment.getFileName(), segment);
+        segment.close();
+    }
+
+    void sync(boolean force) throws IOException {
+        if (force) {
+            for (FileChannel ch : channelsToSync) {
+                ch.force(true);
+                channelsToSync.remove(ch);
+            }
+        }
+        log.trace("Sync'd {} channels", channelsToSync.size());
```

Comment:
This is called often, maybe wrap it with `log.isTrace()`

Reply:
I think no need for that, `channelsToSync` contains a few channels, `size` always fast